exercise-analysis-runner/README.md

```diff
@@ -0,0 +1,265 @@
+# Analysis-runner
+
+As explained in other docs [Getting Started #Analysis-runner](../getting_started.md#analysis-runner), and [Reproducible analysis](../reproducible_analyses.md), the analysis-runner is the gate keeper of running analysis at the CPG.
+
+We don't allow users to run scripts directly on production data, we require code to be commited, reviewed, and then ran. The analysis-runner is the mechanism we use for that indrection. You can sort of think of it like a HPC job manager (like Slurm / PBS).
+
+When we say analysis-runner, we mean the process of an analyst through the CLI, submitting a request for analysis to be run, the AR server accepts the request, authorises that the user can request the analysis, constructs a Batch pipeline from a specific commit in a GitHub repository, and submits as the requested user.
+
+In this exercise, we'll run some existing code through the analysis-runner, dig through the steps and processes - modify the code and run it again.
+
+## Preconditions
+
+You'll need to:
+
+- have git + GitHub set up
+- be added to the `populationgenomics` GitHub organisation
+- ensure an account with Hail Batch has been created
+- be added to the fewgenomes analysis group (speak to your manager if you're unsure about this one)
+- have the analysis-runner CLI tool installed
+
+    - we recommend creating a virtual environment (see [Python](../python.md) for setting up python, and a virtual env trick), and installing the `analysis-runner`.
+
+        ```shell
+        virtualenv analysis-runner
+        pip install analysis-runner
+
+        analysis-runner --version
+        # analysis-runner v2.43.3
+        ```
+
+## Running code via the analysis-runner
+
+We're going to run the `examples/batch/hello.py` Hail Batch workflow via the analysis-runner. This code example is conveniently located in the analysis-runner github repository.
+
+```shell
+git clone git@github.com:populationgenomics/analysis-runner.git
+cd analysis-runner
+```
+
+### The batch to run
+
+Let's take a look at the example:
+
+```python
+from shlex import quote
+
+import click
+from cpg_utils.hail_batch import get_batch
+
+
+@click.command()
+@click.option('--name-to-print')
+def main(name_to_print):
+    """Runs test hail batch workflow"""
+
+    b = get_batch()
+
+    j1 = b.new_job('first job')
+    # For Hail batch, j.{identifier} will create a Resource (file)
+    # that will be collected at the end of a job
+    stdout_of_j = j1.out
+    string_to_print = f'Hello, {name_to_print}'
+    j1.command(f'echo {quote(string_to_print)} > {stdout_of_j}')
+
+    j2 = b.new_job('second job')
+    # for the second job, using an f-string with the resource file
+    # will tell batch to run j2 AFTER j1
+    j2.command(f'cat {stdout_of_j}')
```

Comment:
```suggestion
    string_to_print = f'Hello, {name_to_print}'
    j1.command(f'echo {quote(string_to_print)} > {j1.out}')

    j2 = b.new_job('second job')
    # for the second job, using an f-string with the resource file
    # will tell batch to run j2 AFTER j1
    j2.command(f'cat {j1.out}')
```

I don't feel strongly about this, but assigning `j1.out` to a named variable would make it harder for me to grok the simplicity of this. In all the pipeline code we use `j1.out` in all the format strings, and it may be useful to know that it's not essential to create the `j1.NAME` in advance of sticking it in a String

Reply:
This is the code as it is in the example - so I'm a little hesitant to change it here. I kind of wanted to demonstrate that you use exactly the same reference in both examples, rather than this magic `j.out`, which tbh I don't really like.